daskdev-sat/dask_kubernetes_start.py

```diff
@@ -0,0 +1,146 @@
+    ), "[Dask Cluster] Critical error: could not create worker config"
+    with open(worker_config) as f:
+        d = yaml.safe_load(f)
+        pod_template = make_pod_from_dict(d)
+
+    print(pod_template)  # a lot of interesting undocumented stuff there
+
+    cluster = KubeCluster(
+        pod_template=pod_template,
+        deploy_mode="remote",
+        name=name,
+        namespace=namespace,
+        scheduler_timeout="%d m" % scheduler_timeout_min,
+    )
+
+    if starting_workers > 2:
+        print("Starting scaled up to %s workers" % starting_workers)
+        cluster.scale_up(starting_workers)  # specify number of nodes explicitly
+
+    cluster.adapt(
+        minimum=starting_workers, maximum=max_workers
+    )  # dynamically scale based on current workload
+    return cluster
+
+
+## command line interface
+@click.group()
+@click.pass_context
+def cli(ctx):
+    if ctx.invoked_subcommand is None:
+        print("the command you need is:  'kube-cluster'")
+
+
+"""  Starts infinite async loop for KubeCluster. 
+     It will start scheduler pod and worker pods.     
+"""
+
+
+@cli.command()
+@click.argument("name")  ## cluster name
+@click.option("--namespace")  ## dask cluster namespace
+@click.option("--worker-config")  ## worker config filename
+@click.option("--starting-workers")
+@click.option("--max-workers")
+@click.option("--cpus-per-pod")
+@click.option("--scheduler-timeout-min")
+@click.option("--use-kube-config", default=False)
+def kube_cluster(
+    name,
+    namespace,
+    worker_config,
+    starting_workers,
+    max_workers,
+    cpus_per_pod,
+    scheduler_timeout_min,
+    use_kube_config,
+):
+    loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(loop)
+
+    kube_config_file = "~/.kube/config" if use_kube_config else None
+    if kube_config_file is not None:
+        print("Using kube config file:", kube_config_file)
+
+    params = {"name": name}
+
+    if namespace is not None:
+        params["namespace"] = namespace
+    if worker_config is not None:
+        params["worker_config"] = worker_config
+    if starting_workers is not None:
+        params["starting_workers"] = starting_workers
+    if max_workers is not None:
+        params["max_workers"] = max_workers
+    if cpus_per_pod is not None:
+        params["cpus_per_pod"] = cpus_per_pod
+    if kube_config_file is not None:
+        params["kube_config_file"] = kube_config_file
+    if scheduler_timeout_min is not None:
+        params["scheduler_timeout_min"] = scheduler_timeout_min
+
+    result = loop.run_until_complete(start_dask_kube_cluster(**params))
+    print(f"\nDask cluster {name} started in namespace {namespace}")
+    print(f"\tworker config: {worker_config}")
+    print(f"\tworkers: start: {starting_workers}, max: {max_workers}")
+    print(f"\tcpus per pod {cpus_per_pod}")
+    print(f"\tscheduler timeout: {scheduler_timeout_min} min")
+    print(f"\tscheduler address: {result.scheduler_address}")
```

Comment:
Should these be print or log?

Reply:
You see it saying kubectl logs - and it shows console outputs.  I actually know for sure that if I use -u switch nothing is buffered and I get the latest stuff.   I am pretty sure logging is well buffered, so I'm not so sure we have chosen the right path today..  We need pure immediate console output.  Does logging comply with -u?

Reply:
I found this: (using StreamHandler):
https://stackoverflow.com/questions/55593017/how-to-turn-off-logging-buffer-to-get-logs-in-real-time-with-python-command-line

Reply:
this should definitely be a logging call.  I think the only legitimate reason to use print is if stdout/stdin is part of the UI.  in this case there is no UI (since we're the ones dispatching this container), only logs.  

logging complies with how logging is configured.  If we're not getting the log message the way we want, then we're not configuring it correctly.  In this case, probably calling `logging.config.basicConfig(level=logging.INFO)` should be enough.

Reply:
I did some experimenting.  There is an ultimate thing:  .flush() and that removes all concerns about buffering.

Reply:
I'm pretty sure you just need to configure logging.  If logging isn't properly configured, even if you flush the buffers nothing will be emitted (for example, if you set the log level to ERROR, none of your info statements will print)

Reply:
Configuring log level is an easy part.  The tricky one is to be able to see the last print when do kubectl logs ...   I *had* this not long ago when debugging kubcluster - I knew that it did the print() but I did not see it in the logs (was upsetting).   Had to add -u flag for python - that cured that buffering. How it is going to be with logging - we'll see.  I suspect console output will work the same way (line buffered) and it's fine.  If not - we'll add a Handler and flush() it.